Init statistics sensor upon HASS start

ehendrix23 opened the pull request:
## Description:

Register the state listener and read from the recorder only once HASS has been started.

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

ehendrix23 committed:
Update query to include maxAge

Updated the query from recorded to include MaxAge if set; reducing the amount of records retrieved that would otherwise be purged anyway for the sensor.

ehendrix23 committed:
Initialization upon HASS start

Register the state listener and read previous information from recorder once HASS is started.

ehendrix23 committed:
Updated test_statistics.py for HASS start

Updated test_statistics.py to start HASS and wait it is completed before running test.

ehendrix23 committed:
Added newline in docstring

Added newline in docstring.

ehendrix23 committed:
Added start of HASS to test_initialize_from_database_with_maxage

Added start of HASS to new test test_initialize_from_database_with_maxage.

MartinHjelmare commented on homeassistant/components/sensor/statistics.py:
```diff
@@ -275,5 +286,7 @@ def _purge_old(self):
         for state in reversed(states):
             self._add_state_to_queue(state)
 
+        self.hass.async_add_job(self.async_update_ha_state, True)
```
Why do we need to add this?

ehendrix23 replied:
Once HASS is started, we register our listener for state changes on the entity we calculate statistics for. 
Then we read from the database (if recorder is one of the components) any previous state data available for that entity.
Once we read it, we call to update our own state and thus do calculations based on what was in the database. This call is to do this. 
Another method (I think) would be to use await in our startup and then when we come back to the update state call.

ehendrix23 commented on homeassistant/components/sensor/statistics.py:
```diff
@@ -111,10 +110,22 @@ def async_stats_sensor_state_listener(entity, old_state, new_state):
+
+            if 'recorder' in self._hass.config.components:
+                # only use the database if it's configured
+                self.hass.async_add_job(self._initialize_from_database)
```
Based on my understanding, because we're calling another method that is async (changed the name of this method now to _async_initialize_from_database) it should thus be async_create_task. If it would be a sync method then I would use async_add_executor_job correct?

ehendrix23 commented on homeassistant/components/sensor/statistics.py:
```diff
@@ -111,10 +110,22 @@ def async_stats_sensor_state_listener(entity, old_state, new_state):
             self._add_state_to_queue(new_state)
 
-            hass.async_add_job(self.async_update_ha_state, True)
+            self.hass.async_add_job(self.async_update_ha_state, True)
```
I changed it to self.async_schedule_update_ha_state passing True so that our update method is called first.

ehendrix23 committed:
Updates based on review

Following updates based on review:
-) Removed self._hass and passing hass
-) Changed async_add_job to async_create_task
-) For state update, calling async_schedule_update_ha_state

MartinHjelmare reviewed (approved):
Looks good! Can be merged when build passes.
